File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    answer = 0
    nodes = [[] for i in range(n+1)]
    depths = [0, 1]
    
    for e in edge:
        nodes[e[0]].append(e[1])
        nodes[e[1]].append(e[0])
    visited = [0 for i in range(n+1)]
    visited[1] = 1

    def bfs():
        queue = deque([(1, 0)])
        while queue:
            num, depth = queue.popleft()
            
            print(num, depth)
            if depth > depths[0]:
                depths[0] = depth
                depths[1] = 1
            elif depth == depths[0]:
                depths[1] += 1
                
            for node in nodes[num]:
                if not visited[node]:
                    queue.append((node, depth+1))
                    visited[node] = 1
    bfs()    
    print(depths)  

    return depths[1]

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;
import java.io.*;

class Solution {
    public int solution(int n, int[][] edge) {
        int answer = 0;
        int maxLength = 0;
        int maxCount = 0;
        List<Integer>[] list = new ArrayList[n + 1];
        boolean[] visited = new boolean[n + 1];
        
        for (int i = 0; i <= n; i++) {
            list[i] = new ArrayList();
        }
        
        for (int[] e: edge) {
            list[e[0]].add(e[1]);
            list[e[1]].add(e[0]);
        }

        Queue<int[]> queue = new ArrayDeque<>();
        queue.add(new int[] {1, 0});
        visited[1] = true;
        
        while (!queue.isEmpty()) {
            int[] target = queue.poll();
            
            for (Integer adj: list[target[0]]) {
                if (visited[adj] == true) continue; 
                int length = target[1] + 1;
                if (length  > maxLength) {
                    maxLength = length;
                    maxCount = 1;
                } else if (length == maxLength) {
                    maxCount++;
                }
                queue.add(new int[] {adj, target[1] + 1});
                visited[adj] = true;
            }
        }
        return maxCount;
    }
}